test/UnitTests/Microsoft.Testing.Platform.UnitTests/Filters/FiltersTests.cs

```diff
@@ -0,0 +1,254 @@
+    }
+
+    [TestMethod]
+    public async Task Zero_Available_Filters_Builds_Nop_Filter()
+    {
+        ITestExecutionFilter testExecutionFilter = await GetBuiltFilter(testHost =>
+        {
+            testHost.AddTestExecutionFilter(_ => new Filter1
+            {
+                IsEnabled = false,
+            });
+            testHost.AddTestExecutionFilter(_ => new Filter2
+            {
+                IsEnabled = false,
+            });
+        });
+
+        Assert.IsTrue(testExecutionFilter is NopFilter);
+    }
+
+    [TestMethod]
+    public async Task Single_Registered_Filter_Builds_Single_Filter()
+    {
+        ITestExecutionFilter testExecutionFilter = await GetBuiltFilter(testHost =>
+            testHost.AddTestExecutionFilter(_ => new Filter1()));
+
+        Assert.IsTrue(testExecutionFilter is Filter1);
+    }
+
+    [TestMethod]
+    public async Task Single_Available_Filter_Builds_Single_Filter()
+    {
+        ITestExecutionFilter testExecutionFilter = await GetBuiltFilter(testHost =>
+        {
+            testHost.AddTestExecutionFilter(_ => new Filter1
+            {
+                IsEnabled = false,
+            });
+            testHost.AddTestExecutionFilter(_ => new Filter2());
+        });
+
+        Assert.IsTrue(testExecutionFilter is Filter2);
+    }
+
+    [TestMethod]
+    public async Task Two_Registered_Filters_Builds_Aggregate_Filter()
+    {
+        ITestExecutionFilter testExecutionFilter = await GetBuiltFilter(testHost =>
+        {
+            testHost.AddTestExecutionFilter(_ => new Filter1());
+            testHost.AddTestExecutionFilter(_ => new Filter2());
+        });
+
+        Assert.IsTrue(testExecutionFilter is AggregateFilter);
+        Assert.IsTrue(((AggregateFilter)testExecutionFilter).InnerFilters[0] is Filter1);
+        Assert.IsTrue(((AggregateFilter)testExecutionFilter).InnerFilters[1] is Filter2);
+    }
+
+    [TestMethod]
+    public async Task AggregateFilter_CallsInner_IReceivesAllTestNodesExtension_When_Called_Itself()
+    {
+        ITestExecutionFilter testExecutionFilter = await GetBuiltFilter(testHost =>
+        {
+            testHost.AddTestExecutionFilter(_ => new ReceivableTestNodesFilter());
+            testHost.AddTestExecutionFilter(_ => new ReceivableTestNodesFilter());
+            testHost.AddTestExecutionFilter(_ => new ReceivableTestNodesFilter());
+            testHost.AddTestExecutionFilter(_ => new ReceivableTestNodesFilter());
+            testHost.AddTestExecutionFilter(_ => new ReceivableTestNodesFilter());
+        });
+
+        Assert.IsTrue(testExecutionFilter is AggregateFilter);
+
+        ((AggregateFilter)testExecutionFilter).ReceiveTestNodes([]);
+
+        Assert.IsTrue(((AggregateFilter)testExecutionFilter).InnerFilters.Cast<ReceivableTestNodesFilter>().All(x => x.ReceivedTestNodes));
+    }
+
+    [TestMethod]
+    public async Task Batch_ReceivesAllTestNodes_Filter_Example()
```

Comment:
@Youssef1313 Here's an example of how I'd like to to use `IReceivesAllTestNodesExtension` within a filter.

Reply:
I assume in real-world, you will do something like `if ([run|discover]TestExecutionRequest.Filter is IReceivesAllTestNodesExtension receivesTestNodeFilter) receivesTestNodeFilter.ReceiveTestNodes()`?

Isn't that achievable if you define the interface on your side, and have the special handling for `AggregateFilter`? Something like:

```csharp
if (request.Filter is AggregateFilter aggregateFilter)
{
    foreach (var inner in aggregateFilter.InnerFilters.OfType<IReceivesAllTestNodes>())
    {
        inner.ReceiveTestNodes();
    }
}

(request.Filter as IReceivesAllTestNodes)?.ReceiveTestNodes();
```

Reply:
Yeah you're right actually. I can just define this my side. I'll remove it from this PR!